### PR TITLE
KOGITO-1877: Cucumber tests should allow to change registry/namespace…

### PR DESCRIPTION
**Pull request by Sgitario**
JIRA Ticket: https://issues.redhat.com/browse/KOGITO-1877
Description: 
Added new attributes to allow share the images when running the pipelines:

- build_image_registry
- build_image_namespace
- services_image_registry
- services_image_namespace

We will remove the defaults in the services and images parameters (registry, namespace and version). If not set, the operator will deal with it using the defaults.

Many thanks for submiting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](CONTRIBUTING.MD#sending-a-pull-request)
- [x] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket
- [x] Your feature/bug fix has a unit test that verifies it
- [x] You've tested the new feature/bug fix in an actual OpenShift cluster

### CODE REVIEW

**Comment by Sgitario** [comment]
I've tested a few combinations with these changes:

- For build_image_registry and build_image_namespace:

- [x] without these new parameters. Neither the image runtime nor image S2I were set.
- [x] with build_image_registry only
- [x] with build_image_namespace only
- [x] with both build_image_registry and build_image_namespace

- For services_image_registry and services_image_namespace:

- [x] without these new parameters. Defaults were set.
- [x] with services_image_registry only
- [x] with services_image_namespace only
- [x] with both services_image_registry and services_image_namespace

File: pkg/infrastructure/kogito_images.go [hunk 1]
```diff
@@ -32,6 +37,11 @@ func GetRuntimeImageVersion() string {
 	return getRuntimeImageVersion(version.Version)
 }
 
+// GetDefaultImage returns the default full image of the image name.
+func GetDefaultImage(imageName string) string {
```
Comment (sutaakar) [inline pkg/infrastructure/kogito_images.go:41]: A nitpick, can you please move `GetDefaultImage` up so `GetRuntimeImageVersion()` and `getRuntimeImageVersion(v string)` are next to each other, as they are related?

File: pkg/infrastructure/services/deployment_test.go [hunk 2]
```diff
@@ -15,14 +15,15 @@
...
 	"github.com/kiegroup/kogito-cloud-operator/pkg/apis/app/v1alpha1"
 	"github.com/kiegroup/kogito-cloud-operator/pkg/infrastructure"
 	"github.com/stretchr/testify/assert"
-	"k8s.io/apimachinery/pkg/apis/meta/v1"
-	"testing"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-const defaultDataIndexImageFullTag = infrastructure.DefaultDataIndexImageNoVersion + "latest"
+const defaultDataIndexImageFullTag = "quay.io/kiegroup/" + infrastructure.DefaultDataIndexImageName + ":latest"
```
Comment (sutaakar) [inline pkg/infrastructure/services/deployment_test.go:26]:
```suggestion
var defaultDataIndexImageFullTag = infrastructure.GetDataIndexServiceDefaultImage()
```

File: test/steps/kogitoapp.go [hunk 3]
```diff
@@ -30,7 +30,7 @@ func registerKogitoAppSteps(s *godog.Suite, data *Data) {
 	s.Step(`^Deploy (quarkus|springboot) example service "([^"]*)" with configuration:$`, data.deployExampleServiceWithConfiguration)
 	s.Step(`^Create (quarkus|springboot) service "([^"]*)"$`, data.createService)
 	s.Step(`^Create (quarkus|springboot) service "([^"]*)" with configuration:$`, data.createServiceWithConfiguration)
-	s.Step(`^Deploy service from example file "([^"]*)"$`, data.deployServiceFromExampleFile)
+	s.Step(`^Deploy (quarkus|springboot) service from example file "([^"]*)"$`, data.deployServiceFromExampleFile)
```
Comment (sutaakar) [inline test/steps/kogitoapp.go:33]: Could you please move this change into separate PR?
It doesn't seem to be related to the JIRA.

File: pkg/infrastructure/services/deployment_test.go (hunk 2 above)
Comment (sutaakar) [inline pkg/infrastructure/services/deployment_test.go:26]:
```suggestion
var defaultDataIndexImageFullTag = infrastructure.GetDataIndexServiceDefaultImage()
```

File: test/steps/kogitoapp.go (hunk 3 above)
Comment (Sgitario) [reply test/steps/kogitoapp.go:33]: I had to make this change in here since we need to set the runtimeType now in the constructor of the KogitoAppStub, so we can resolve the images to use accordingly based on the build type and the runtime type.

File: pkg/infrastructure/services/deployment_test.go (hunk 2 above)
Comment (Sgitario) [reply pkg/infrastructure/services/deployment_test.go:26]: This is not the same. Runtime version will return 0.10.0 instead of latest and I didn't want to modify this test... What do you think?

File: pkg/infrastructure/kogito_images.go (hunk 1 above)
Comment (Sgitario) [reply pkg/infrastructure/kogito_images.go:41]: Well spotted!

File: pkg/infrastructure/services/deployment_test.go (hunk 2 above)
Comment (sutaakar) [reply pkg/infrastructure/services/deployment_test.go:26]: The tests here don't check the Data index tag itself, it is used just as a parameter for triggering the tested method.
The change above doesn't affect the tests.

File: test/steps/kogitoapp.go (hunk 3 above)
Comment (sutaakar) [reply test/steps/kogitoapp.go:33]: Oh, ok, I have overlooked that.

File: pkg/infrastructure/services/deployment_test.go (hunk 2 above)
Comment (Sgitario) [reply pkg/infrastructure/services/deployment_test.go:26]: ok ok. Will change this then.

File: test/steps/kogitoapp.go (hunk 3 above)
Comment (Sgitario) [reply test/steps/kogitoapp.go:33]: Also, this step is not used anywhere, but I thought it was better to maintain it just in case you're creating your own scenarios for testing purposes.

File: test/framework/kogitoapp.go [hunk 4]
```diff
@@ -211,6 +214,42 @@ func setupBuildImageStreams(kogitoApp *v1alpha1.KogitoApp) {
...
+	if len(config.GetBuildImageRegistry()) > 0 || len(config.GetBuildImageNamespace()) > 0 {
+		overwriteBuildImageRuntimeIfNotSet(kogitoApp)
+		overwriteBuildImageS2IIfNotSet(kogitoApp)
+	}
+
 	// If "KOGITO_BUILD_IMAGE_VERSION" is defined, it's taken into account, otherwise set the current version
 	kogitoApp.Spec.Build.ImageVersion = config.GetBuildImageVersion()
 }
+
+func overwriteBuildImageRuntimeIfNotSet(kogitoApp *v1alpha1.KogitoApp) {
```
Comment (sutaakar) [inline test/framework/kogitoapp.go:226]: What about adjusting this function to something like
```suggestion
func getBuildRuntimeImageStreamTag(kogitoApp *v1alpha1.KogitoApp) string {
```
And here you would completely resolve the tag?
The logic will stay the same, just in case` len(config.GetBuildRuntimeImageStreamTag()) > 0` you can return `config.GetBuildRuntimeImageStreamTag()`.
This would IMHO be more readable.

File: test/steps/kogitoapp.go [hunk 5]
```diff
@@ -65,7 +63,7 @@ func (data *Data) createService(runtimeType, serviceName string) error {
 }
 
 func (data *Data) createServiceWithConfiguration(runtimeType, serviceName string, table *messages.PickleStepArgument_PickleTable) error {
-	kogitoApp := framework.GetKogitoAppStub(data.Namespace, serviceName)
+	kogitoApp := framework.GetKogitoAppStub(data.Namespace, runtimeType, serviceName)
 	if err := configureKogitoAppFromTable(table, kogitoApp); err != nil {
 		return err
 	}
```
Comment (sutaakar) [inline test/steps/kogitoapp.go:69]: Please remove `kogitoApp.Spec.Runtime = v1alpha1.RuntimeType(runtimeType)` on line 71 as it is not needed any more.

Comment (Sgitario) [reply test/steps/kogitoapp.go:69]: ops, sorry. Removed.

File: test/framework/kogitoapp.go (hunk 4 above)
Comment (Sgitario) [reply test/framework/kogitoapp.go]: I refactor this change to try making more readable. Let me know your thoughts now.
Thanks!

File: test/framework/kogitoapp.go [hunk 6]
```diff
@@ -209,8 +212,56 @@ func setupBuildImageStreams(kogitoApp *v1alpha1.KogitoApp) {
...
+
+	return ""
+}
+
+func isBuildImageRegistryOrNamespaceSet() bool {
+	return len(config.GetBuildImageRegistry()) > 0 || len(config.GetBuildImageNamespace()) > 0
+}
+
+func getBuildImage(image string) string {
+	registry := config.GetBuildImageRegistry()
```
Comment (radtriste) [inline test/framework/kogitoapp.go:256]: why not using the `Image` struc from operator framework package and use the convert method at the end ?

File: test/config/config.go [hunk 7]
```diff
@@ -104,13 +109,17 @@ func BindFlags(set *flag.FlagSet) {
...
+	set.StringVar(&env.servicesImageRegistry, prefix+"services-image-registry", infrastructure.DefaultImageRegistry, "Set the services (jobs-service, data-index) image registry")
 	set.StringVar(&env.dataIndexImageTag, prefix+"data-index-image-tag", "", "Set the Kogito Data Index image tag ('services-image-version' is ignored)")
 	set.StringVar(&env.jobsServiceImageTag, prefix+"jobs-service-image-tag", "", "Set the Kogito Jobs Service image tag ('services-image-version' is ignored)")
 	set.StringVar(&env.mgmtConsoleImageTag, prefix+"management-console-image-tag", "", "Set the Kogito Management Console image tag ('services-image-version' is ignored)")
 
 	// build
 	set.StringVar(&env.mavenMirrorURL, prefix+"maven-mirror-url", "", "Maven mirror url to be used when building app in the tests")
 	set.StringVar(&env.buildImageVersion, prefix+"build-image-version", infrastructure.GetRuntimeImageVersion(), "Set the build image version")
+	set.StringVar(&env.buildImageNamespace, prefix+"build-image-namespace", "", "Set the build image namespace")
```
Comment (radtriste) [inline test/config/config.go:121]: I find it confusing to have default values for services but not for build ...

File: test/framework/kogitoapp.go [hunk 8]
```diff
@@ -209,8 +212,56 @@ func setupBuildImageStreams(kogitoApp *v1alpha1.KogitoApp) {
...
+	if len(registry) == 0 {
```
Comment (radtriste) [inline test/framework/kogitoapp.go:257]: having default values for build, would allow us to avoid this check

File: test/framework/kogitoapp.go [hunk 9]
```diff
@@ -209,8 +212,56 @@ func setupBuildImageStreams(kogitoApp *v1alpha1.KogitoApp) {
...
+		registry = infrastructure.DefaultImageRegistry
+	}
+
+	namespace := config.GetBuildImageNamespace()
+	if len(namespace) == 0 {
```
Comment (radtriste) [inline test/framework/kogitoapp.go:262]: same here for the check

File: test/config/config.go (hunk 7 above)
Comment (Sgitario) [reply test/config/config.go:121]: I did this way at first, but then I realise that with default values, we'll always be sending the image runtime and image S2I attributes, and then we will stop covering the ability of the operator to set these values.

Also, I didn't want to change how was done at first. With services, we always send the image values. That's why services have defaults and images not. 

What do you think about it? I still can change it..

File: test/framework/kogitoapp.go (hunk 8 above)
Comment (Sgitario) [reply test/framework/kogitoapp.go:257]: Read my reply above.

File: test/framework/kogitoapp.go (hunk 9 above)
Comment (Sgitario) [reply test/framework/kogitoapp.go:262]: Read my reply above.

File: test/framework/kogitoapp.go (hunk 6 above)
Comment (Sgitario) [reply test/framework/kogitoapp.go:256]: Do you mean, to use this method framework.ConvertImageToImageTag(image), right? Will change it.

File: test/config/config.go (hunk 7 above)
Comment (radtriste) [reply test/config/config.go:121]: > With services, we always send the image values

what do you mean ?
In the code of tests, we always force the services image ?

> and then we will stop covering the ability of the operator to set these values.

This is the same for services in fact.

Indeed, we are setting most of the time the build and services.
But we should have same behavior for build and for services I think.

IMO, we should have default values for image version/namespace/registry and set only if given.
In that case, we can still execute the tests letting the operator set the default values.
wdyt ?

File: test/framework/kogitoapp.go (hunk 6 above)
Comment (radtriste) [reply test/framework/kogitoapp.go:256]: yes that is right.
That would avoid any mistake when building the image name :)

Comment (Sgitario) [reply test/framework/kogitoapp.go:256]: Changed

File: test/config/config.go (hunk 7 above)
Comment (Sgitario) [reply test/config/config.go:119]: As agreed, I removed the defaults in the services and images parameters (registry, namespace and version). If not set, the operator will deal with it using the defaults, so we can cover this functionality in the operator.

File: test/framework/kogitoserviceutils.go [hunk 10]
```diff
@@ -122,7 +145,11 @@ func cliInstall(service v1alpha1.KogitoService, cliName string, cliFlags []strin
 		cmd = append(cmd, cliFlag)
 	}
 
-	cmd = append(cmd, "--image", framework.ConvertImageToImageTag(*service.GetSpec().GetImage()))
+	image := framework.ConvertImageToImageTag(*service.GetSpec().GetImage())
```
Comment (Sgitario) [inline test/framework/kogitoserviceutils.go:148]: This change was not necessary, but it was weird to send only "--image" with a value. In fact, The operator is behaving differently when sending "--image" (without value) and when not sending the "--image". I raised a defect for this: https://issues.redhat.com/browse/KOGITO-1961

File: test/framework/kogitoapp.go [hunk 11]
```diff
@@ -209,8 +213,66 @@ func setupBuildImageStreams(kogitoApp *v1alpha1.KogitoApp) {
...
+		namespace = infrastructure.DefaultImageNamespace
+	}
+
+	version := config.GetBuildImageVersion()
+	if len(version) == 0 {
+		version = infrastructure.GetRuntimeImageVersion()
+	}
+
+	return framework.ConvertImageToImageTag(v1alpha1.Image{
```
Comment (radtriste) [inline test/framework/kogitoapp.go:272]: why don't you create the image at the beginning of the method and update the fields directly instead of creating variables ?

File: test/framework/kogitoserviceutils.go [hunk 12]
```diff
@@ -98,14 +99,36 @@ func GetCliFlags(persistence, events bool) []string {
...
+func newImageOrDefault(fullImage string, defaultImageName string) v1alpha1.Image {
 	if len(fullImage) > 0 {
 		return framework.ConvertImageTagToImage(fullImage)
 	}
 
-	image := framework.ConvertImageTagToImage(defaultImage)
-	image.Tag = config.GetServicesImageVersion()
-	return image
+	if len(config.GetServicesImageRegistry()) > 0 || len(config.GetServicesImageNamespace()) > 0 || len(config.GetServicesImageVersion()) > 0 {
+		registry := config.GetServicesImageRegistry()
```
Comment (radtriste) [inline test/framework/kogitoserviceutils.go:108]: same here you, you could create the image with data from config and update the fields if empty ?

File: test/framework/kogitoserviceutils.go (hunk 10 above)
Comment (radtriste) [reply test/framework/kogitoserviceutils.go:148]: good for me

File: test/framework/kogitoapp.go (hunk 11 above)
Comment (Sgitario) [reply test/framework/kogitoapp.go:272]: Okis, changed.

File: test/framework/kogitoserviceutils.go (hunk 12 above)
Comment (Sgitario) [reply test/framework/kogitoserviceutils.go:108]: Okis, changed.

File: pkg/infrastructure/kogito_images.go [hunk 13]
```diff
@@ -16,16 +16,26 @@ package infrastructure
...
 	LatestTag = "latest"
+	// DefaultImageRegistry the default services image repository
+	DefaultImageRegistry = "quay.io"
+	// DefaultImageNamespace the default services image namespace
+	DefaultImageNamespace = "kiegroup"
 )
 
+// GetDefaultImage returns the default full image of the image name.
+func GetDefaultImage(imageName string) string {
+	return DefaultImageRegistry + "/" + DefaultImageNamespace + "/" + imageName + ":" + GetRuntimeImageVersion()
```
Comment (ricardozanini) [inline pkg/infrastructure/kogito_images.go:36]: Please avoid this kind of concatenation on functions body, try this:

```suggestion
	return fmt.sprintf( "%s/%s/%s:%s", DefaultImageRegistry, DefaultImageNamespace, imageName, GetRuntimeImageVersion())
```
Test it before commiting since I wrote from the top of my head. xD

Comment (Sgitario) [reply pkg/infrastructure/kogito_images.go]: Thanks for spotting. I missed to update this piece by using:

```
return framework.ConvertImageToImageTag(v1alpha1.Image{
		Domain:    DefaultImageRegistry,
		Namespace: DefaultImageNamespace,
		Name:      imageName,
		Tag:       GetRuntimeImageVersion(),
	})
```

**Comment by Sgitario** [comment]
Rebased and tested again.
